node/impl/full/eth.go

```diff
@@ -1258,10 +1263,47 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
 		return nil, api.ErrNotSupported
 	}
 
+	if e.EventFilterManager.EventIndex == nil {
+		return nil, xerrors.Errorf("cannot use eth_get_logs if historical event index is disabled")
+	}
+
+	pf, err := e.parseEthFilterSpec(ctx, filterSpec)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to parse eth filter spec: %w", err)
+	}
+
+	if pf.tipsetCid == cid.Undef {
+		maxHeight := pf.maxHeight
+		if maxHeight == -1 {
+			maxHeight = e.Chain.GetHeaviestTipSet().Height()
+		}
+		err := e.waitForHeightProcessed(ctx, maxHeight)
```

Comment:
I kind of assumed the event indexing system would handle this for us. Although the install/take/uninstall pattern makes that hard.

Maybe add a `GetEvents` method to the `EventFilterManager` that waits if necessary then returns the events (and doesn't require the install/uninstall dance)?

Reply:
Yeah, we should _really_ do it this way. You'll be able to drop all the subscribe stuff as well.

Basically:

1. Add a `sync.Cond` to `EventFilterManager`.
2. First up, check to make sure the latest tipset we have in the chainstore is beyond the requested tipset/height.
3. Take `EventFilterManager.mu`
4. Compare `currentHeight` against the target height.
5. If `currentHeight` is less than the target, wait on the condition and repeat step 4.
6. If greater than or equal to, check if we've actually processed the tipset we care about. If not, abort.
7. Release the lock and collect the relevant events.

Reply:
Hm. A cond variable is going to be hard if we want to obey the context.

But... I think there's a better way. We can:

1. Have event filters decide if they're done (as they process events).
2. Add a `WaitFinished` (or something) method to event filters that waits for them to declare that they're "done" (can match no more events.

Reply:
I would like to address this as part of https://github.com/filecoin-project/lotus/issues/12116 rather than solving this in a patchy way here. For now, this PR does solve the problem of `eth_getLogs` failing to return events for tipsets on the cannonical chain and I would like to ship this as an intermediate step since we're now anyways going to prioritise https://github.com/filecoin-project/lotus/issues/12116

Reply:
My proposal should be much simpler than the current subscribe logic, unless I'm missing something. This kind of subscribe pattern tends to have deadlock issues (although I haven't looked at your implementation closely beyond suggesting an alternative).

Reply:
Nvm. We discussed this and @aarshkshah1992's proposal is better overall.